src/utils/queue.ts

```diff
@@ -1,41 +1,55 @@
-export type Task<T> = () => Promise<T>
-
-export default class Mutex {
-  private queue: Array<Task<any>> = []
-  private isBusy = false
-  private limit: number
-  private completed: Promise<any> = Promise.resolve()
-
-  constructor (limit: number = Infinity) {
-    this.limit = limit
-  }
-
-  async runExclusive<T> (task: Task<T>): Promise<T> {
-    // Only add another item to the queue if it's below the limit
-    if (this.limit >= this.queue.length + 1) {
-      this.completed = new Promise<T>(async (resolve, reject) => {
-        this.queue.push(() => task().then(resolve).catch(reject))
+enum QueueState {
+  /** Not running any tasks; immediately ready to start a new task */
+  IDLE,
+  /** Running a task */
+  BUSY
+}
 
-        if (!this.isBusy) {
-          await this.dequeue()
-        }
+export default class Queue {
+  private state = QueueState.IDLE
+  private queue: Array<{
+    task: () => Promise<any>,
+    priority: number
+  }> = []
+
+  /**
+   * Add a new task to the queue with the given priority
+   * @param priority greater numbers represent tasks that will run sooner than lesser numbers
```

Comment:
I would add a comment about what the expected behavior is for tasks with the same priority. FIFO, LIFO, no guarantee? Not clear to me at least.